qlib/workflow/online/manager.py

```diff
@@ -75,82 +87,60 @@ def first_train(self, strategies: List[OnlineStrategy] = None, model_kwargs: dic
             models = self.trainer.train(tasks, experiment_name=strategy.name_id)
-            models_list.append(models)
+            models = self.trainer.end_train(models, experiment_name=strategy.name_id)
+            self.logger.info(f"Finished training {len(models)} models.")
 
-        for strategy, models in zip(strategies, models_list):
-            self.prepare_online_models(strategy, models, model_kwargs=model_kwargs)
+            online_models = strategy.prepare_online_models(models, **model_kwargs)
+            self.history.setdefault(self.cur_time, {})[strategy] = online_models
 
     def routine(
         self,
         cur_time: Union[str, pd.Timestamp] = None,
-        delay: bool = False,
         task_kwargs: dict = {},
         model_kwargs: dict = {},
         signal_kwargs: dict = {},
     ):
         """
-        Run typical update process for every strategy and record the online history.
+        Typical update process for every strategy and record the online history.
 
         The typical update process after a routine, such as day by day or month by month.
-        The process is: Prepare signals -> Prepare tasks -> Prepare online models.
+        The process is: Update predictions -> Prepare tasks -> Prepare online models -> Prepare signals.
 
         If using DelayTrainer, it can finish training all together after every strategy's prepare_tasks.
 
         Args:
             cur_time (Union[str,pd.Timestamp], optional): run routine method in this time. Defaults to None.
-            delay (bool): if delay prepare signals and models
             task_kwargs (dict): the params for `prepare_tasks`
             model_kwargs (dict): the params for `prepare_online_models`
             signal_kwargs (dict): the params for `prepare_signals`
         """
         if cur_time is None:
             cur_time = D.calendar(freq=self.freq).max()
         self.cur_time = pd.Timestamp(cur_time)  # None for latest date
-        models_list = []
-        for strategy in self.strategy:
-            if not delay:
-                strategy.tool.update_online_pred()
+
+        for strategy in self.strategies:
             self.logger.info(f"Strategy `{strategy.name_id}` begins routine...")
+            if self.status == self.STATUS_NORMAL:
+                strategy.tool.update_online_pred()
 
             tasks = strategy.prepare_tasks(self.cur_time, **task_kwargs)
             models = self.trainer.train(tasks)
-            models_list.append(models)
+            if self.status == self.STATUS_NORMAL or not self.trainer.is_delay():
+                models = self.trainer.end_train(models, experiment_name=strategy.name_id)
+            self.logger.info(f"Finished training {len(models)} models.")
+            online_models = strategy.prepare_online_models(models, **model_kwargs)
+            self.history.setdefault(self.cur_time, {})[strategy] = online_models
 
-        if not delay:
+        if not self.trainer.is_delay():
```

Comment:
It will better to add a 2*2 table in this file to explain how `status` and `delay` affect the behavior of the online manager.